c++/src/kj/compat/http.c++

```diff
@@ -5380,6 +6085,50 @@ private:
     return kj::heap<BrokenWebSocket>(KJ_EXCEPTION(FAILED,
         "received bad WebSocket handshake", errorMessage));
   }
+
+  kj::Own<kj::AsyncIoStream> accept(uint statusCode, kj::StringPtr statusText,
+                                    const HttpHeaders& headers) override {
+    auto method = KJ_REQUIRE_NONNULL(currentMethod, "already called send()");
+    currentMethod = nullptr;
+    KJ_ASSERT(method.is<HttpConnectMethod>(), "only use accept() with CONNECT requests");
+
+    KJ_REQUIRE(statusCode >= 200 && statusCode < 300, "the statusCode must be 2xx for accept");
+    tunnelRejected = nullptr;
+
+    httpOutput.writeHeaders(headers.serializeResponse(statusCode, statusText));
+
+    auto releasedBuffer = httpInput.releaseBuffer();
+    auto deferNoteClosed = kj::defer([this]() { webSocketOrConnectClosed = true; });
+    kj::Own<kj::AsyncIoStream> ownStream(&stream, kj::NullDisposer::instance);
+
+    auto promised = httpOutput.flush().then([io = ownStream.attach(kj::mv(deferNoteClosed))]
+        () mutable -> kj::Promise<kj::Own<kj::AsyncIoStream>> {
+      return kj::mv(io);
+    });
+
+    return heap<AsyncIoStreamWithInitialBuffer>(
+        kj::newPromisedStream(kj::mv(promised)),
+        kj::mv(releasedBuffer.buffer),
+        releasedBuffer.leftover);
+  }
+
+  kj::Own<kj::AsyncOutputStream> reject(uint statusCode, kj::StringPtr statusText,
+                           const HttpHeaders& headers) override {
+    auto method = KJ_REQUIRE_NONNULL(currentMethod, "already called send()");
+    currentMethod = nullptr;
+    KJ_REQUIRE(method.is<HttpConnectMethod>(), "Only use reject() with CONNECT requests.");
+
+    KJ_REQUIRE(statusCode < 200 || statusCode >= 300, "the statusCode must not be 2xx for reject.");
+    tunnelRejected = Maybe<kj::Promise<bool>>(true);
+
+    // We are always going to send a Connection: Close with rejections, just to be safe.
+    kj::StringPtr connectionHeaders[HttpHeaders::CONNECTION_HEADERS_COUNT];
+    connectionHeaders[HttpHeaders::BuiltinIndices::CONNECTION] = "close";
+
+    httpOutput.writeHeaders(headers.serializeResponse(
+        statusCode, statusText, connectionHeaders));
+    return heap<HttpPassthroughWriter>(httpOutput);
```

Comment:
I think that `reject()` should take `kj::Maybe<uint64_t> expectedBodySize` like `send()` does, and construct the appropriate kind of body stream. Even though we're going to close the connection anyway, it's still polite to delimit the body as we would normally.